### ISSUE TITLE
Python API for iterating over variants  

### ISSUE THREAD

**Issue by percyfal** [sgkit-dev/vcztools#112]
Having briefly glossed through the code, my understanding is that vcztools will target similar functionality as bcftools, converting Zarr to VCF. Would other output formats fit in this tool or should that end up elsewhere? I'm thinking of `cyvcf2.cyvcf2.Variant` output as this would allow downstream applications that rely on `cyvcf2` to directly access Zarr archives through a Python API. The use case I'm thinking of right now is ancestral allele reconstruction with fastDFE for tsinfer analyses (see https://github.com/Sendrowski/fastDFE/issues/13). I guess it would be better for vcztools to focus mainly on bcftools functionality and that other file format outputs end up in a separate toolkit?

**Comment by jeromekelleher** [comment]
I think it's out of scope for vcztools, as we're trying to keep this tightly focused on emulating bcftools.

What you suggest is a great idea though - I guess the natural place for it is in cyvcf2? It would be a bit of work, and we'd need to convince Brent that it was worthwhile

**Comment by percyfal** [comment]
Yeah, I figured as much. I wonder if it would be out of scope for cyvcf2 too though as that is a cython wrapper around htslib and expects VCF/BCF as input? Maybe we need a zarr2bio package :sweat_smile: ?

**Comment by jeromekelleher** [comment]
Yeah, it would depend on Brent's enthusiasm a lot. 

I guess the fundamental thing you need to do is efficiently iterate over the arrays in variant order, and select by position. We could imagine breaking this out into its own package, which vcztools and other python libs could import?

**Comment by jeromekelleher** [comment]
Yeah, it would depend on Brent's enthusiasm a lot. 

I guess the fundamental thing you need to do is efficiently iterate over the arrays in variant order, and select by position. We could imagine breaking this out into its own package, which vcztools and other python libs could import?

We could also just export that iteration API from this package I guess...

**Comment by percyfal** [comment]
I guess that was one of my first thoughts; if there is an exposable Python API in vcztools that iterates over variants, other libraries could easily build on that to convert to whatever. vcztools could still focus on bcftools compliance, where the API is a convenience for other libraries.

**Comment by jeromekelleher** [comment]
Yes, I think you're right. This is a non trivial thing that would be really useful to have availabile, particularly with background fetching of chunks and caching etc. Great idea!

**Comment by jeromekelleher** [comment]
I guess we could offer a duck-type best effort layer on top of this that aims to behave like cyvcf2? It would be fiddly, as the API is a bit quirky and could be a moving target.

**Comment by percyfal** [comment]
So to conclude, if I start playing around with this, should I a) draft an API in a separate branch of vcztools b) do the conversion to cyvcf2 in an auxiliary library (personal at the moment)?

**Comment by jeromekelleher** [comment]
Let's draft in vcztools please, once there's no additional dependencies I'm happy to add it here.

**Comment by tomwhite** [comment]
This should be fairly straightforward now. The following is a quick illustration of how to iterate over variants, with support for filtering by regions, filter expressions, and samples.

```python
import zarr

from vcztools import retrieval
from vcztools.samples import parse_samples


def variant_iterator(
    vcz,
    *,
    fields=None,
    regions=None,
    targets=None,
    samples=None,
    include=None,
    exclude=None,
):
    root = zarr.open(vcz, mode="r")
    all_samples = root["sample_id"][:]
    _, samples_selection = parse_samples(samples, all_samples)

    for chunk_data in retrieval.variant_chunk_iter(
        root,
        fields=fields,
        variant_regions=regions,
        variant_targets=targets,
        include=include,
        exclude=exclude,
        samples_selection=samples_selection,
    ):
        pos = chunk_data["variant_position"]
        num_variants = len(pos)
        for i in range(num_variants):
            yield i, chunk_data


if __name__ == "__main__":
    for i, chunk_data in variant_iterator(
        "vcz_test_cache/sample.vcf.vcz", fields=["variant_position", "call_genotype"]
    ):
        pos = chunk_data["variant_position"][i]
        gt = chunk_data["call_genotype"][i]
        print(f"{i}\t{pos}\t{tuple(g.tolist() for g in gt)}")
```

Output:

```
0	111	([0, 0], [0, 0], [0, 1])
1	112	([0, 0], [0, 0], [0, 1])
2	14370	([0, 0], [1, 0], [1, 1])
3	17330	([0, 0], [0, 1], [0, 0])
4	1110696	([1, 2], [2, 1], [2, 2])
5	1230237	([0, 0], [0, 0], [0, 0])
6	1234567	([0, 1], [0, 2], [-1, -1])
7	1235237	([0, 0], [0, 0], [-1, -1])
8	10	([0, -2], [0, 1], [0, 2])
```

Perhaps it would be better to subset all of the arrays in `chunk_data` to a single row and just yield that? (May be a perf trade off in constructing a new dict - or maybe not - needs measuring.)
